src/CADL.Extension/Emitter.Csharp/src/lib/model.ts

```diff
@@ -407,27 +349,121 @@ export function getInputType(
         }
     }
 
-    function getInputTypeForArray(
-        program: Program,
-        elementType: Type
-    ): InputListType {
+    function getInputTypeForArray(elementType: Type): InputListType {
         return {
             Name: "Array",
             ElementType: getInputType(program, elementType, models, enums),
             IsNullable: false
         } as InputListType;
     }
 
-    function getInputTypeForMap(
-        program: Program,
-        key: Type,
-        value: Type
-    ): InputType {
+    function getInputTypeForMap(key: Type, value: Type): InputDictionaryType {
         return {
             Name: "Dictionary",
             KeyType: getInputType(program, key, models, enums),
             ValueType: getInputType(program, value, models, enums),
             IsNullable: false
         } as InputDictionaryType;
+    }   
+
+    function getInputModelForModel(m: ModelType): InputModelType {
+        m = getEffectiveSchemaType(program, m) as ModelType;
+        const name = getFriendlyName(program, m) ?? m.name;
+        let model = models.get(name);
+        if (!model) {
+            const baseModel = getInputModelBaseType(m.baseModel);
+            const properties: InputModelProperty[] = [];
+
+            model = {
+                Name: name,
+                Namespace: m.namespace?.name,
+                Description: getDoc(program, m),
+                IsNullable: false,
+                DiscriminatorPropertyName: getDiscriminator(program, m)?.propertyName,
+                DiscriminatorValue: getDiscriminatorValue(baseModel),
```

Comment:
DiscriminatorValue is defined in derived model, not the base Model.

Reply:
Yes, it picks up model from the outer scope, but I'll change the signature to make it easier to read

Reply:
Fixed